### PR TITLE
Porting STM32 Ethernet HAL driver

### LINKED ISSUE #41380: stm32h7: Ethernet: Migrate driver to the new eth HAL api
**Is your enhancement proposal related to a problem? Please describe.**
A new eth cube hal api is being pushed in STM32Cube (https://github.com/STMicroelectronics/stm32h7xx_hal_driver/pull/16).
It is first deployed to H7 series and will be then be progressively extended to other STM32 series.
This new HAL provides significant performance benefits and additional features (cf related github PR)
To use it, existing STM32 ethernet HAL driver should be reworked.

I'm raising this issue to synchronize efforts on this task.

EDIT: A STM32H7Cube package has been proposed that includes examples using the new HAL API. It's available here:
https://github.com/STMicroelectronics/STM32CubeH7/pull/195

### PR DESCRIPTION
**Pull request by gleichdick**
This is a follow-up of #45260 to fix #41380.

A new Kconfig option (<del>`ETH_STM32_USE_HAL_DRIVER_WITH_NEW_API`</del>`ETH_STM32_HAL_API_V2`) is introduced to switch to the non-legacy ETH HAL driver. Furthermore, multiple receive buffers can be used for one single packet, which means that the rx buffer size can be reduced and doesn't have to match the maximum ethernet packet length anymore.

Future work:
 - [ ] Make rx buffer size selectable by config option (?)
 - [x] Support multiple buffers for TX path as well for the same reason
 - [ ] Use new PTP HAL

Depends on zephyrproject-rtos/hal_stm32#139

### CODE REVIEW

File: drivers/ethernet/Kconfig.stm32_hal [hunk 1]
```diff
@@ -152,4 +152,11 @@ config ETH_STM32_HAL_PTP_CLOCK_INIT_PRIO
 
 endif # SOC_SERIES_STM32F7X || SOC_SERIES_STM32H7X
 
+config ETH_STM32_USE_HAL_DRIVER_WITH_NEW_API
+	bool "Use new HAL driver"
+	default no
```
Comment (bbilas) [inline drivers/ethernet/Kconfig.stm32_hal:157]: That is not needed, please remove it. Please see https://docs.zephyrproject.org/1.14.0/guides/kconfig/index.html#redundant-defaults

Comment (bbilas) [inline drivers/ethernet/Kconfig.stm32_hal:157]: That is not needed, please remove it. Please see https://docs.zephyrproject.org/1.14.0/guides/kconfig/index.html#redundant-defaults

**Comment by erwango** [comment]
@gleichdick Thanks for this updated version!
In order to ease the use of your PR, please add a commit which will update manifest to use the modified hal_stm32:
In `west.yml`, do:
```
    - name: hal_stm32
-      revision: 6253094a4147cbd30d2b466c975eb4013ac6be90
+     revision: pull/139/head
      path: modules/hal/stm32
      groups:
        - hal
```

File: drivers/ethernet/Kconfig.stm32_hal [hunk 2]
```diff
@@ -152,4 +152,11 @@ config ETH_STM32_HAL_PTP_CLOCK_INIT_PRIO
 
 endif # SOC_SERIES_STM32F7X || SOC_SERIES_STM32H7X
 
+config ETH_STM32_USE_HAL_DRIVER_WITH_NEW_API
```
[the remark this answers is not in the document]
Comment (erwango) [reply drivers/ethernet/Kconfig.stm32_hal:155]: Also, as this is a somewhat important option for the short term future, please put it much higher in the file (eg line 17).
Also potentially there will be impact on the subsequent options, so better to declare it before them.

File: drivers/ethernet/eth_stm32_hal.c [hunk 3]
```diff
@@ -439,7 +562,18 @@ static struct net_pkt *eth_rx(const struct device *dev, uint16_t *vlan_tag)
 
 	heth = &dev_data->heth;
 
-#if defined(CONFIG_SOC_SERIES_STM32H7X)
+#if defined(CONFIG_ETH_STM32_HAL_API_V2)
+	if (HAL_ETH_ReadData(heth, &appbuf) != HAL_OK) {
+		/* no frame available */
+		return NULL;
+	}
+
```
Comment (gleichdick) [inline drivers/ethernet/eth_stm32_hal.c:570]: By looking in the disassembly, it seems that gcc 10.3.1 optimizes away the case that a packet has been received. I suspect that this is due to some undefined behaviour (UB). Compiling with `-fno-strict-aliasing` does not change anything, but with `-Og` it seems to be fine (as `eth_rx` is not inlined)

Comment (gleichdick) [inline drivers/ethernet/eth_stm32_hal.c:570]: By looking in the disassembly, it seems that gcc 10.3.1 optimizes away the case that a packet has been received. I suspect that this is due to some undefined behaviour (UB). Compiling with `-fno-strict-aliasing` does not change anything, but with `-Og` it seems to be fine (as `eth_rx` is not inlined)

**Review by gleichdick — COMMENTED** [review]
I'm finally able to run some tests on a Nucleo H743. The new driver seems to work in general (tested the telnet example), but suffers some initial packet losses <del>and does not work at all with `-O2` enabled</del>.
```
$  ping  stm32
PING stm32 (192.0.2.1) 56(84) bytes of data.
64 bytes from stm32 (192.0.2.1): icmp_seq=9 ttl=64 time=1.09 ms
64 bytes from stm32 (192.0.2.1): icmp_seq=10 ttl=64 time=0.577 ms
64 bytes from stm32 (192.0.2.1): icmp_seq=11 ttl=64 time=0.517 ms
64 bytes from stm32 (192.0.2.1): icmp_seq=12 ttl=64 time=0.452 ms
64 bytes from stm32 (192.0.2.1): icmp_seq=13 ttl=64 time=0.511 ms
```
First 8 ping packages are lost

File: drivers/ethernet/eth_stm32_hal.c (hunk 3 above)
Comment (gleichdick) [reply drivers/ethernet/eth_stm32_hal.c:570]: Actually, this is not true. `eth_rx` is inlined but nothing is optimized away

**Comment by gleichdick** [comment]
The package loss was due to the arp cache. No package losses when clearing the host arp cache after mcu reset.

**Comment by erwango** [comment]
@gleichdick Is this ready for review ?

**Comment by erwango** [comment]
@gleichdick Please let us know the status for this work. Thanks!

**Comment by gleichdick** [comment]
Yes, I think this ready now. Thank you.

File: drivers/ethernet/eth_stm32_hal.c [hunk 4]
```diff
@@ -89,7 +89,27 @@ static ETH_DMADescTypeDef dma_tx_desc_tab[ETH_TXBUFNB] __eth_stm32_desc;
...
+struct eth_stm32_tx_context {
+	struct net_pkt *pkt;
+	uint16_t tx_buffer_index;
+};
+
+static struct eth_stm32_rx_buffer_header dma_rx_buffer_header[ETH_RXBUFNB];
+static bool dma_tx_buffer_used[ETH_TXBUFNB];
+
+#endif /* defined(CONFIG_ETH_STM32_HAL_API_V2) */
```
Comment (erwango) [inline drivers/ethernet/eth_stm32_hal.c:191]: Nitpick:
```suggestion
#endif /* CONFIG_ETH_STM32_HAL_API_V2 */
```
Apply everywhere

File: drivers/ethernet/eth_stm32_hal.c [hunk 5]
```diff
@@ -270,41 +274,61 @@ static int eth_tx(const struct device *dev, struct net_pkt *pkt)
...
-
 	total_len = net_pkt_get_len(pkt);
 	if (total_len > ETH_TX_BUF_SIZE) {
 		LOG_ERR("PKT too big");
-		res = -EIO;
-		goto error;
+		return -EIO;
+	}
+
+	k_mutex_lock(&dev_data->tx_mutex, K_FOREVER);
```
Comment (erwango) [inline drivers/ethernet/eth_stm32_hal.c:318]: Changes to existing code should be done in a dedcaited commit with justification. Ideally a dedicated PR if there are multiple.

File: drivers/ethernet/eth_stm32_hal.c [hunk 6]
```diff
@@ -284,18 +319,8 @@ static int eth_tx(const struct device *dev, struct net_pkt *pkt)
...
-				goto buffer_found;
-			}
-		}
-		ctx.tx_buffer_index = 0;
-		k_yield();
-	}
-
-buffer_found:
+	ctx.first_tx_buffer_index = allocate_tx_buffer();
+	buf_header = &dma_tx_buffer_header[ctx.first_tx_buffer_index];
```
Comment (erwango) [inline drivers/ethernet/eth_stm32_hal.c:323]: Can you squash this initial commit, so we don't review code that is later deleted ?

File: drivers/ethernet/eth_stm32_hal.c [hunk 7]
```diff
@@ -191,40 +309,73 @@ static int eth_tx(const struct device *dev, struct net_pkt *pkt)
...
 	total_len = net_pkt_get_len(pkt);
-	if (total_len > ETH_TX_BUF_SIZE) {
+	if (total_len > (ETH_TX_BUF_SIZE * ETH_TXBUFNB)) {
 		LOG_ERR("PKT too big");
-		res = -EIO;
-		goto error;
+		return -EIO;
 	}
 
+	k_mutex_lock(&dev_data->tx_mutex, K_FOREVER);
```
Comment (erwango) [inline drivers/ethernet/eth_stm32_hal.c:318]: Changes to existing code should be done in dedicated commit and explained.

File: drivers/ethernet/eth_stm32_hal.c [hunk 8]
```diff
@@ -89,7 +89,108 @@ static ETH_DMADescTypeDef dma_tx_desc_tab[ETH_TXBUFNB] __eth_stm32_desc;
...
+BUILD_ASSERT(ETH_RX_BUF_SIZE % 4 == 0, "Rx buffer size must be a multiple of 4");
+
+struct eth_stm32_rx_buffer_header {
+	struct eth_stm32_rx_buffer_header *next;
+	uint16_t size;
+	bool used;
+};
+
+struct eth_stm32_tx_buffer_header {
+	ETH_BufferTypeDef hal_def;
```
Comment (erwango) [inline drivers/ethernet/eth_stm32_hal.c:103]: `hal_def` is too generic, what about:
```suggestion
	ETH_BufferTypeDef tx_buff;
```

**Review by erwango — CHANGES_REQUESTED** [review]
Here is a first round of comments.

Also, can you check about commits formating: https://static.sched.com/hosted_files/zephyr2022/a3/HOWTO_Get_Your_Zephyr_Patches_Merged_ZDS_2022.pdf ?

File: drivers/ethernet/eth_stm32_hal.c (hunk 5 above)
Comment (gleichdick) [reply drivers/ethernet/eth_stm32_hal.c]: Here it is: 34898e22d8a7

**Comment by gleichdick** [comment]
I also had an idea to add
```c
#ifdef CONFIG_ETH_STM32_HAL_API_V2
#error "This KConfig option is not fully implemented"
#endif
```
In the first commit and to remove it in the last. Even though the code compiles after each commit (at least it should), the new api only works after the last commit.

File: drivers/ethernet/eth_stm32_hal.c [hunk 9]
```diff
@@ -812,6 +812,10 @@ void HAL_ETH_TxCpltCallback(ETH_HandleTypeDef *heth_handle)
 	k_sem_give(&dev_data->tx_int_sem);
 
 }
+#endif /* CONFIG_SOC_SERIES_STM32H7X || CONFIG_ETH_STM32_HAL_API_V2 */
+
+#if defined(CONFIG_ETH_STM32_HAL_API_V2)
+#elif defined(CONFIG_SOC_SERIES_STM32H7X)
```
Comment (erwango) [inline drivers/ethernet/eth_stm32_hal.c:950]: Let's move this to the next commit for simplicity.

File: drivers/ethernet/eth_stm32_hal.c [hunk 10]
```diff
@@ -199,12 +199,12 @@ static HAL_StatusTypeDef read_eth_phy_register(ETH_HandleTypeDef *heth,
...
 						uint32_t *RegVal)
 {
-#if defined(CONFIG_SOC_SERIES_STM32H7X)
+#if defined(CONFIG_SOC_SERIES_STM32H7X) || defined(CONFIG_ETH_STM32_HAL_API_V2)
 	return HAL_ETH_ReadPHYRegister(heth, PHYAddr, PHYReg, RegVal);
 #else
 	ARG_UNUSED(PHYAddr);
 	return HAL_ETH_ReadPHYRegister(heth, PHYReg, RegVal);
-#endif /* CONFIG_SOC_SERIES_STM32H7X */
+#endif /* CONFIG_SOC_SERIES_STM32H7X || CONFIG_ETH_STM32_HAL_API_V2 */
```
Comment (erwango) [inline drivers/ethernet/eth_stm32_hal.c:207]: Link with the current commit ?

File: drivers/ethernet/eth_stm32_hal.c [hunk 11]
```diff
@@ -815,6 +815,31 @@ void HAL_ETH_TxCpltCallback(ETH_HandleTypeDef *heth_handle)
 #endif /* CONFIG_SOC_SERIES_STM32H7X || CONFIG_ETH_STM32_HAL_API_V2 */
 
 #if defined(CONFIG_ETH_STM32_HAL_API_V2)
+void HAL_ETH_ErrorCallback(ETH_HandleTypeDef *heth)
+{
+	__ASSERT_NO_MSG(heth != NULL);
+
+	const uint32_t error_code = HAL_ETH_GetError(heth);
```
Comment (erwango) [inline drivers/ethernet/eth_stm32_hal.c:929]: Why `const`?

File: drivers/ethernet/eth_stm32_hal.c [hunk 12]
```diff
@@ -304,38 +310,72 @@ static int eth_tx(const struct device *dev, struct net_pkt *pkt)
 	heth = &dev_data->heth;
 
 	total_len = net_pkt_get_len(pkt);
-	if (total_len > ETH_TX_BUF_SIZE) {
+	if (total_len > (ETH_TX_BUF_SIZE * ETH_TXBUFNB)) {
```
Comment (erwango) [inline drivers/ethernet/eth_stm32_hal.c:313]: Move this in a dedicated commit as it impacts existing code

File: drivers/ethernet/eth_stm32_hal.c [hunk 13]
```diff
@@ -153,6 +153,41 @@ void HAL_ETH_RxLinkCallback(void **pStart, void **pEnd, uint8_t *buff, uint16_t
...
+	for (;;) {
+		for (uint16_t index = 0; index < ETH_TXBUFNB; index++) {
+			if (!dma_tx_buffer_header[index].used) {
+				dma_tx_buffer_header[index].used = true;
+				return index;
+			}
+		}
+		k_yield();
+	}
+}
```
Comment (erwango) [inline drivers/ethernet/eth_stm32_hal.c:190]: Seem this function belongs to one of the next commits

File: drivers/ethernet/eth_stm32_hal.c [hunk 14]
```diff
@@ -304,38 +310,72 @@ static int eth_tx(const struct device *dev, struct net_pkt *pkt)
...
 	timestamped_frame = eth_is_ptp_pkt(net_pkt_iface(pkt), pkt);
 	if (timestamped_frame) {
 		/* Enable transmit timestamp */
-#if defined(CONFIG_SOC_SERIES_STM32H7X)
+#if defined(CONFIG_ETH_STM32_HAL_API_V2)
+		HAL_ETH_PTP_InsertTxTimestamp(heth);
+#elif defined(CONFIG_SOC_SERIES_STM32H7X)
 		dma_tx_desc->DESC2 |= ETH_DMATXNDESCRF_TTSE;
-#else
+#else /* CONFIG_SOC_SERIES_STM32H7X */
```
Comment (erwango) [inline drivers/ethernet/eth_stm32_hal.c:348]: The comment on else is confusing

File: drivers/ethernet/eth_stm32_hal.c [hunk 15]
```diff
@@ -347,15 +387,22 @@ static int eth_tx(const struct device *dev, struct net_pkt *pkt)
 		goto error;
 	}
 
-#if defined(CONFIG_SOC_SERIES_STM32H7X)
```
Comment (erwango) [inline drivers/ethernet/eth_stm32_hal.c:239]: Isn't this breaking existing NON H7 code ?

File: drivers/ethernet/eth_stm32_hal.c [hunk 16]
```diff
@@ -696,15 +722,36 @@ static struct net_pkt *eth_rx(const struct device *dev, uint16_t *vlan_tag)
 		goto release_desc;
 	}
 
+#if defined(CONFIG_ETH_STM32_HAL_API_V2)
+	for (rx_header = (struct eth_stm32_rx_buffer_header *)appbuf;
+			rx_header; rx_header = rx_header->next) {
+		const size_t index = rx_header - &dma_rx_buffer_header[0];
```
Comment (erwango) [inline drivers/ethernet/eth_stm32_hal.c:728]: why `const` ?

File: drivers/ethernet/eth_stm32_hal.c [hunk 17]
```diff
@@ -191,40 +309,73 @@ static int eth_tx(const struct device *dev, struct net_pkt *pkt)
...
 		LOG_ERR("PKT too big");
-		res = -EIO;
-		goto error;
+		return -EIO;
 	}
 
+	k_mutex_lock(&dev_data->tx_mutex, K_FOREVER);
+
+#if defined(CONFIG_ETH_STM32_HAL_API_V2)
+
```
Comment (erwango) [inline drivers/ethernet/eth_stm32_hal.c:321]: Nitpick: looking to existing style inside this file, no spaces are needed around these lines.

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for the update.

One additional comment regarding the commit messages.
I suggest to update titles in the following way:
`drivers: ethernet: stm32: bla`

See https://youtu.be/mnVNLM7ynMQ?t=936

File: drivers/ethernet/eth_stm32_hal.c (hunk 15 above)
Comment (gleichdick) [reply drivers/ethernet/eth_stm32_hal.c:239]: Yes, it is. Good catch!

File: drivers/ethernet/eth_stm32_hal.c (hunk 14 above)
Comment (gleichdick) [reply drivers/ethernet/eth_stm32_hal.c:348]: Removed

File: drivers/ethernet/eth_stm32_hal.c (hunk 12 above)
Comment (gleichdick) [reply drivers/ethernet/eth_stm32_hal.c:313]: Done

File: drivers/ethernet/eth_stm32_hal.c (hunk 9 above)
Comment (gleichdick) [reply drivers/ethernet/eth_stm32_hal.c:950]: Okay

File: drivers/ethernet/eth_stm32_hal.c (hunk 10 above)
Comment (gleichdick) [reply drivers/ethernet/eth_stm32_hal.c:207]: Could you elaborate on that?

Comment (erwango) [reply drivers/ethernet/eth_stm32_hal.c:207]: Title of the commit is "drivers: ethernet: stm32: Add transmit callbacks"

Comment (gleichdick) [reply drivers/ethernet/eth_stm32_hal.c:207]: Thanks! Like this?

**Comment by GeorgeCGV** [comment]
> @GeorgeCGV As I recall you're an STM32 eth user. Would you ave time to review and test this ?

I am sorry, but I am not using STM32 ETH.  I am not sure where that idea comes from.

**Comment by erwango** [comment]
> I am sorry, but I am not using STM32 ETH. I am not sure where that idea comes from.

Ok, my bad

**Comment by erwango** [comment]
@gleichdick Would you mind rebasing this ?
I'd like to merge it soon. It would be disabled by default, so no risk, and easier to test.

**Review by erwango — DISMISSED** [review]

**Review by FRASTM — DISMISSED** [review]

**Comment by erwango** [comment]
@gleichdick I'm getting errors when trying to build with new API:

```
$ west build  -b nucleo_h723zg samples/net/dhcpv4_client
```

```
/local/mcu/zephyrproject/zephyr/drivers/ethernet/eth_stm32_hal.c:95:14: error: 'ETH_RX_BUF_SIZE' undeclared here (not in a function); did you mean 'ETH_STM32_RX_BUF_SIZE'?
   95 | BUILD_ASSERT(ETH_RX_BUF_SIZE % 4 == 0, "Rx buffer size must be a multiple of 4");
      |              ^~~~~~~~~~~~~~~
/local/mcu/zephyrproject/zephyr/drivers/ethernet/eth_stm32_hal.c:95:14: error: expression in static assertion is not an integer
/local/mcu/zephyrproject/zephyr/drivers/ethernet/eth_stm32_hal.c: In function 'HAL_ETH_RxLinkCallback':
/local/mcu/zephyrproject/zephyr/drivers/ethernet/eth_stm32_hal.c:139:9: error: type defaults to 'int' in type name [-Werror=implicit-int]
  139 |         size_t index = (RxBufferPtr)buff - &dma_rx_buffer[0];
      |         ^~~~~~
/local/mcu/zephyrproject/zephyr/drivers/ethernet/eth_stm32_hal.c:139:42: error: invalid operands to binary - (have 'int' and 'uint8_t (*)[1528]' {aka 'unsigned char (*)[1528]'})
  139 |         size_t index = (RxBufferPtr)buff - &dma_rx_buffer[0];
      |                        ~~~~~~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~
      |                        |                   |
      |                        int                 uint8_t (*)[1528] {aka unsigned char (*)[1528]}
/local/mcu/zephyrproject/zephyr/drivers/ethernet/eth_stm32_hal.c: In function 'eth_tx':
/local/mcu/zephyrproject/zephyr/drivers/ethernet/eth_stm32_hal.c:356:33: error: 'ETH_TX_BUF_SIZE' undeclared (first use in this function); did you mean 'ETH_STM32_TX_BUF_SIZE'?
  356 |         while (remaining_read > ETH_TX_BUF_SIZE) {
      |                                 ^~~~~~~~~~~~~~~
      |                                 ETH_STM32_TX_BUF_SIZE
/local/mcu/zephyrproject/zephyr/drivers/ethernet/eth_stm32_hal.c:356:33: note: each undeclared identifier is reported only once for each function it appears in

```

**Comment by gleichdick** [comment]
cc9b6bb0b4e removed `ETH_{R,T}X_BUF_SIZE` definitions, I replaced them with `ETH_STM32_{R,T}X_BUF_SIZE`

**Review by erwango — DISMISSED** [review]

**Review by erwango — DISMISSED** [review]

**Review by erwango — APPROVED** [review]

**Review by FRASTM — APPROVED** [review]
